### PR TITLE
feat: Get songs by artist using iTunes API

### PR DESCRIPTION
**Pull request by aseer-ws**
### Ticket Link

---

### Related Links

---

### Description

- Integrate iTunes API into iTunes container
- Render songs by artist

### Steps to Reproduce / Test

---

---

### Checklist

- [ ] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added]
- [ ] Relevant documentation is changed or added (and PR referenced)

### GIF's

TunesApp

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -1 +1,2 @@
-GITHUB_URL=https://api.github.com/
\ No newline at end of file
+GITHUB_URL=https://api.github.com/
+ITUNES_URL=https://itunes.apple.com/
```
Comment (christin-wednesday) [inline .env]: Please add EOF

File: app/components/TrackCard/index.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+function TrackCard({ trackId, trackName, collectionName, trackPrice, releaseDate }) {
+  return (
+    <StyledCol md={12} lg={8} style={{ height: '100%' }}>
+      <CustomCard>
+        <If condition={!isEmpty(trackName)} otherwise={<T id="track_name_unavailable" />}>
+          <T id="track_name" values={{ trackName }} />
+        </If>
+        <p>Collection name: {collectionName}</p>
+        <p></p>
```
Comment (christin-wednesday) [inline app/components/TrackCard/index.js]:
```suggestion
```

File: app/containers/TunesContainer/index.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+  const renderTunesError = () => {
+    return <div>{JSON.stringify(tunesError, null, 2)}</div>;
+  };
+
+  return (
+    <Container maxWidth={maxWidth} padding={padding}>
+      <Helmet>
+        <title>TunesContainer</title>
+        <meta name="description" content="Description of TunesContainer" />
```
Comment (christin-wednesday) [inline app/containers/TunesContainer/index.js:147]: Nice 👍🏽

File: app/containers/TunesContainer/index.js [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      </Helmet>
+      <StyledHeader>
+        <T type="heading" marginBottom={10} id="itunes_header" />
+        <Search
+          placeholder="input search text"
+          allowClear
+          defaultValue={artist}
+          onChange={(evt) => onDebouncedSearch(evt.target.value)}
+          onSearch={(term) => onDebouncedSearch(term)}
+          style={{ width: 304 }}
```
Comment (christin-wednesday) [inline app/containers/TunesContainer/index.js]: It might be better to stick with max-width.

Comment (christin-wednesday) [reply app/containers/TunesContainer/index.js]: Also we could create a StyledComponent for this

File: app/containers/TunesContainer/reducer.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+};
+
+export const { Types: tunesContainerTypes, Creators: tunesContainerCreators } = createActions({
+  setArtist: ['artistName'],
+  setSongsData: ['songsData'],
+  setTunesError: ['error'],
+  clearSongs: {}
+});
+
+/* eslint-disable default-case, no-param-reassign */
```
Comment (christin-wednesday) [inline app/containers/TunesContainer/reducer.js]:
```suggestion
```

File: app/containers/TunesContainer/selectors.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import { initialState } from './reducer';
+import get from 'lodash/get';
+
+/**
+ * Direct selector to the tunesContainer state domain
+ */
+
+const selectTunesContainerDomain = (state) => state.tunesContainer || initialState;
+
+const makeSelectTunesContainer = () => createSelector(selectTunesContainerDomain, (substate) => substate);
```
Comment (christin-wednesday) [inline app/containers/TunesContainer/selectors.js]:
```suggestion
```

We can remove this

File: app/containers/TunesContainer/reducer.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+export const initialState = {
+  artist: '',
+  songsData: {},
+  tunesError: null
+};
+
+export const { Types: tunesContainerTypes, Creators: tunesContainerCreators } = createActions({
+  setArtist: ['artistName'],
+  setSongsData: ['songsData'],
```
Comment (christin-wednesday) [inline app/containers/TunesContainer/reducer.js]: A pattern that we follow for API request action names is

requestX  - action the saga will listen for and which is dispatched from the Container
successX - action dispatched from the saga on success
failureX - action dispatched from the saga on success

File: app/containers/TunesContainer/tests/reducer.test.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  let state;
+  beforeEach(() => {
+    state = initialState;
+  });
+
+  it('should return the initial state', () => {
+    expect(tunesContainerReducer(undefined, {})).toEqual(state);
+  });
+
+  it('should return the update the state when an action of type DEFAULT is dispatched', () => {
```
Comment (christin-wednesday) [inline app/containers/TunesContainer/tests/reducer.test.js]: We could also update these test descriptions

**Review by christin-wednesday — COMMENTED** [review]
Please add Screenshots.

File: app/components/TrackCard/index.js [hunk 9]
```diff
@@ -1,45 +1,88 @@
...
   && {
-    min-height: 100px;
     height: 100%;
   }
 `;
 
-function TrackCard({ trackId, trackName, collectionName, trackPrice, releaseDate }) {
+const StyledImage = styled(Image)`
+  && {
+    width: 100px;
```
Comment (christin-wednesday) [inline app/components/TrackCard/index.js]: please consider rem for dimensions.

File: app/components/TrackCard/index.js [hunk 10]
```diff
@@ -1,45 +1,88 @@
...
...dv22uIEiLA4q7nvvCug8WTqzQveOH26fodo7g6uFe/a17W3+nFBAkRYENRdb1vkkz1CH9cPsVy/jrhr27PqMYvENYNlHAIesRiBYwRy0V+8iXP8+/fvX11Mr7L7ECueb/r48eMqm7FuI2BGWDEG8cm+7G3NEOfmdcTQw4h9/55lhm7DekRYKQPZF2ArbXTAyu4kDYB2YxUzwg0gi/41ztHnfQG26HbGel/crVrm7tNY+/1btkOEAZ2M05r4FB7r9GbAIdxaZYrHdOsgJ/wCEQY0J74TmOKnbxxT9n3FgGGWWsVdowHtjt9Nnvf7yQM2aZU/TIAIAxrw6dOnAWtZZcoEnBpNuTuObWMEiLAx1HY0ZQJEmHJ3HNvGCBBhY6jtaMoEiJB0Z29vL6ls58vxPcO8/zfrdo5qvKO+d3Fx8Wu8zf1dW4p/cPzLly/dtv9Ts/EbcvGAHhHyfBIhZ6NSiIBTo0LNNtScABFyNiqFCBChULMNNSdAhJyNSiECRCjUbEPNCRAhZ6NSiAARCjXbUHMCRMjZqBQiQIRCzTbUnAARcjYqhQgQoVCzDTUnQIScjUohAkQo1GxDzQkQIWejUogAEQo121BzAkTI2agUIkCEQs021JwAEXI2KoUIEKFQsw01J0CEnI1KIQJEKNRsQ80JECFno1KIABEKNdtQcwJEyNmoFCJAhELNNtScABFyNiqFCBChULMNNSdAhJyNSiECRCjUbEPNCRAhZ6NSiAARCjXbUHMCRMjZqBQiQIRCzTbUnAARcjYqhQgQoVCzDTUnQIScjUohAkQo1GxDzQkQIWejUogAEQo121BzAkTI2agUIkCEQs021JwAEXI2KoUIEKFQsw01J0CEnI1KIQJEKNRsQ80JECFno1KIABEKNdtQcwJEyNmoFCJAhELNNtScABFyNiqFCBChULMNNSdAhJyNSiECRCjUbEPNCRAhZ6NSiAARCjXbUHMCRMjZqBQiQIRCzTbUnAARcjYqhQgQoVCzDTUnQIScjUohAkQo1GxDzQkQIWejUogAEQo121BzAkTI2agUIkCEQs021JwAEXI2KoUIEKFQsw01J0CEnI1KIQJEKNRsQ80JECFno1KIABEKNdtQcwJEyNmoFCJAhELNNtScABFyNiqFCBChULMNNSdAhJyNSiEC/wGgKKC4YMA4TAAAAABJRU5ErkJggg=="
```
Comment (christin-wednesday) [inline app/components/TrackCard/index.js]: why is this in this format ?
can we pass in a static image import as the fallback ?

File: app/containers/HomeContainer/tests/reducer.test.js [hunk 11]
```diff
@@ -11,7 +11,7 @@ describe('HomContainer reducer tests', () => {
     expect(homeContainerReducer(undefined, {})).toEqual(state);
   });
 
-  it('should return the initial state when an action of type FETCH_USER is dispatched', () => {
+  it('should return the initial state when an action of type REQUEST_GET?_GITHUB_REPOS is dispatched', () => {
```
Comment (christin-wednesday) [inline app/containers/HomeContainer/tests/reducer.test.js]:
```suggestion
  it('should return the initial state when an action of type REQUEST_GET_GITHUB_REPOS is dispatched', () => {
```

File: app/containers/TunesContainer/index.js [hunk 12]
```diff
@@ -144,15 +155,15 @@ TunesContainer.defaultProps = {
 };
 
 const mapStateToProps = createStructuredSelector({
-  tunesContainer: makeSelectTunesContainer(),
+  tunesContainer: selectTunesContainer(),
```
Comment (christin-wednesday) [inline app/containers/TunesContainer/index.js]: Are we using this prop ?

Comment (aseer-ws) [reply app/containers/TunesContainer/index.js]: yes. in a test

Comment (christin-wednesday) [reply app/containers/TunesContainer/index.js]: Since we have individual selectors, let us remove this selector and it's tests.

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -29,12 +29,12 @@ jobs:
...
-      - name: SonarCloud Scan
-        uses: sonarsource/sonarcloud-github-action@master
-        with:
-          args: >
-            -Dsonar.organization=${{ secrets.SONAR_ORG}}
-            -Dsonar.projectKey=${{ secrets.SONAR_PROJECT_KEY}}
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+      # - name: SonarCloud Scan
```
Comment (christin-wednesday) [inline .github/workflows/ci.yml]: Let us remove this

File: app/components/TrackCard/index.js [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+            </Col>
+          </Row>
+        </StyledTrackCard>
+      </StyledTrackItem>
+    </Skeleton>
+  );
+}
+
+TrackCard.propTypes = {
+  skeletonLoading: PropTypes.bool,
```
Comment (christin-wednesday) [inline app/components/TrackCard/index.js:85]: Are there any of these prop-types required ?

File: app/containers/HomeContainer/tests/reducer.test.js [hunk 15]
```diff
@@ -11,7 +11,7 @@ describe('HomContainer reducer tests', () => {
     expect(homeContainerReducer(undefined, {})).toEqual(state);
   });
 
-  it('should return the initial state when an action of type FETCH_USER is dispatched', () => {
+  it('should return the initial state when an action of type REQUEST_GET_GITHUB_REPOS is dispatched', () => {
```
Comment (christin-wednesday) [inline app/containers/HomeContainer/tests/reducer.test.js:14]: Please revert these changes

File: app/containers/TunesContainer/index.js [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+
+  const onArtistSearch = (artistName) => {
+    if (!artistName) {
+      return dispatchClearSongs();
+    }
+    setLoading(true);
+    dispatchGetArtistSongs(artistName);
+  };
+
+  const onDebouncedSearch = debounce(onArtistSearch, 500);
```
Comment (christin-wednesday) [inline app/containers/TunesContainer/index.js]:
```suggestion
  const handleDebouncedSearch = debounce(onArtistSearch, 500);
```

File: app/jsconfig.json [hunk 17]
```diff
@@ -3,6 +3,7 @@
     "target": "es2017",
     "allowSyntheticDefaultImports": false,
     "baseUrl": "./",
+    "moduleResolution": "node",
```
Comment (christin-wednesday) [inline app/jsconfig.json:6]: Are you able to peek definitions after this change ?

File: app/services/tests/tunesApi.test.js [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+  it('should make api call to /search?term=', async () => {
+    const mock = new MockAdapter(getApiClient('itunes').axiosInstance);
+    const data = {
+      resultCount: 1,
+      results: [{ artistName }]
+    };
+
+    mock.onGet(`search?term=${artistName}`).reply(200, data);
+    const res = await getTunes(artistName);
+    expect(res.data).toEqual(data);
```
Comment (christin-wednesday) [inline app/services/tests/tunesApi.test.js:22]: Should we add a test when `artistName` is undefined ?

File: app/utils/routeConstants.js [hunk 19]
```diff
@@ -1,10 +1,18 @@
 export default {
   repos: {
-    route: '/',
+    route: '/repos',
```
Comment (christin-wednesday) [inline app/utils/routeConstants.js:3]: shouldn't this be some other route ?

File: app/containers/TunesContainer/tests/index.test.js [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+      resultCount: 1,
+      results: [{ artistName }]
+    };
+    const { getByTestId } = renderProvider(
+      <TunesContainer artist={artistName} songsData={songsData} dispatchClearSongs={submitSpy} />
+    );
+
+    fireEvent.change(getByTestId(artistSearchBarId), { target: { value: '' } });
+
+    await timeout(500);
```
Comment (christin-wednesday) [inline app/containers/TunesContainer/tests/index.test.js:65]: why do we need this timeout ?

File: app/containers/TunesContainer/tests/index.test.js [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+  it('should trigger dispatchGetArtistSongs when a change is made in search bar', async () => {
+    const { getByTestId } = renderProvider(<TunesContainer dispatchGetArtistSongs={submitSpy} />);
+
+    fireEvent.change(getByTestId(artistSearchBarId), {
+      target: {
+        value: 'A'
+      }
+    });
+
+    await timeout(500);
```
Comment (christin-wednesday) [inline app/containers/TunesContainer/tests/index.test.js:40]: Will this test fail if we remove this timeout ?

File: app/components/TrackCard/index.js (hunk 14 above)
Comment (aseer-ws) [reply app/components/TrackCard/index.js:85]: no

File: app/jsconfig.json (hunk 17 above)
Comment (aseer-ws) [reply app/jsconfig.json:6]: yup

File: app/containers/TunesContainer/tests/index.test.js (hunk 20 above)
Comment (aseer-ws) [reply app/containers/TunesContainer/tests/index.test.js:65]: because search bar onChange takes in debounced submitSpy

File: app/containers/TunesContainer/tests/index.test.js (hunk 21 above)
Comment (aseer-ws) [reply app/containers/TunesContainer/tests/index.test.js:40]: Screenshot 2022-02-02 at 18 59 08

Comment (aseer-ws) [reply app/containers/TunesContainer/tests/index.test.js:40]: the submitSpy won't get called immediately

File: app/utils/routeConstants.js (hunk 19 above)
Comment (aseer-ws) [reply app/utils/routeConstants.js:3]: I put `TunesContainer` in `/` route and `HomeContainer` in `/repos` route

**Review by christin-wednesday — APPROVED** [review]
